pep-9999.rst

```diff
@@ -0,0 +1,393 @@
+PEP: 9999
+Title: Type defaults for TypeVars
+Version: $Revision$
+Last-Modified: $Date$
+Author: James Hilton-Balfe <gobot1234yt@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To: typing-sig@python.org
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 14-Jul-2022
+Python-Version: 3.12
+
+Type defaults for TypeVars
+==========================
+
+Abstract
+--------
+
+This PEP introduces the concept of type defaults for TypeVars, which act
+as defaults for a type parameter when none is specified.
+
+Motivation
+----------
+
+.. code:: py
+
+   T = TypeVar("T", default=int)  # This means that if no type is specified T = int
+
+   @dataclass
+   class Box(Generic[T]):
+       value: T | None = None
+
+   reveal_type(Box())                      # type is Box[int]
+   reveal_type(Box(value="Hello World!"))  # type is Box[str]
+
+One place this `regularly comes
+up <https://github.com/python/typing/issues/975>`__ is ``Generator``. I
+propose changing the *stub definition* to something like:
+
+.. code:: py
+
+   YieldT = TypeVar("YieldT")
+   SendT = TypeVar("SendT", default=None)
+   ReturnT = TypeVar("ReturnT", default=None)
+
+   class Generator(Generic[YieldT, SendT, ReturnT]): ...
+
+   Generator[int] == Generator[int, None] == Generator[int, None, None]
+
+This is also useful for a Generic that is commonly over one type.
+
+.. code:: py
+
+   class Bot: ...
+
+   BotT = TypeVar("BotT", bound=Bot, default=Bot)
+
+   class Context(Generic[BotT]):
+       bot: BotT
+
+   class MyBot(Bot): ...
+
+   reveal_type(Context().bot)         # type is Bot  # notice this is not Any which is what it would be currently
+   reveal_type(Context[MyBot]().bot)  # type is MyBot
+
+Not only does this improve typing for those who explicitly use it. It
+also helps non-typing users who rely on auto-complete to speed up their
+development.
+
+This design pattern is common in projects like:
+ - `discord.py <https://github.com/Rapptz/discord.py>`__ - where the
+   example above was taken from.
+ - `NumPy <https://github.com/numpy/numpy>`__ - the default for types
+   like ``ndarray``\ 's ``dtype`` would be ``float64``. Currently it's
+   ``Unknown`` or ``Any``.
+ - `TensorFlow <https://github.com/tensorflow/tensorflow>`__ (this
+   could be used for Tensor similarly to ``numpy.ndarray`` and would be
+   useful to simplify the definition of ``Layer``)
+
+`This proposal could also be used on builtins.slice <https://github.com/python/typing/issues/159>`__
+where the parameter should ``start`` default to ``int``, ``stop``
+default to ``start`` and step default to ``int | None``
```

Comment:
```suggestion
where the parameter ``start`` should default to ``int``, ``stop`` shouldd
default to the type of ``start`` and ``step`` should default to ``int | None``
```
Also, do you mean `stop` should also default to `int`?

Reply:
No I thought for convince sake they'd be the same type most of the time so it'd save people the extra type 99%  of the time